File: sdks/java/src/main/java/dev/atap/sdk/crypto/DPoPProof.java
package dev.atap.sdk.crypto;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.UUID;

/**
 * DPoP proof JWT generation (RFC 9449) for Ed25519.
 */
public final class DPoPProof {

    private DPoPProof() {
    }

    /**
     * Create a DPoP proof JWT.
     *
     * @param keyPair     Ed25519 keypair for signing
     * @param method      HTTP method (GET, POST, etc.)
     * @param url         Full URL for the htu claim (must use https://{platformDomain}/path)
     * @param accessToken if provided, includes ath (access token hash) claim
     * @return compact JWS string (header.payload.signature)
     */
    public static String create(Ed25519KeyPair keyPair, String method, String url, String accessToken) {
        String x = keyPair.getPublicKeyBase64Url();

        // Build header JSON
        String header = "{\"alg\":\"EdDSA\",\"jwk\":{\"crv\":\"Ed25519\",\"kty\":\"OKP\",\"x\":\""
                + x + "\"},\"typ\":\"dpop+jwt\"}";

        // Build payload JSON
        StringBuilder payload = new StringBuilder();
        payload.append("{");

        if (accessToken != null && !accessToken.isEmpty()) {
            try {
                MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
                byte[] athBytes = sha256.digest(accessToken.getBytes(StandardCharsets.US_ASCII));
                payload.append("\"ath\":\"").append(Base64Url.encode(athBytes)).append("\",");
            } catch (NoSuchAlgorithmException e) {
                throw new RuntimeException("SHA-256 not available", e);
            }
        }

        long iat = System.currentTimeMillis() / 1000;
        String jti = UUID.randomUUID().toString();

        payload.append("\"htm\":\"").append(method).append("\",");
        payload.append("\"htu\":\"").append(url).append("\",");
        payload.append("\"iat\":").append(iat).append(",");
        payload.append("\"jti\":\"").append(jti).append("\"");
        payload.append("}");

        String headerB64 = Base64Url.encode(header.getBytes(StandardCharsets.UTF_8));
        String payloadB64 = Base64Url.encode(payload.toString().getBytes(StandardCharsets.UTF_8));

        byte[] signingInput = (headerB64 + "." + payloadB64).getBytes(StandardCharsets.US_ASCII);
        byte[] signature = keyPair.sign(signingInput);

        return headerB64 + "." + payloadB64 + "." + Base64Url.encode(signature);
    }

    /**
     * Create a DPoP proof JWT without access token hash.
     */
    public static String create(Ed25519KeyPair keyPair, String method, String url) {
        return create(keyPair, method, url, null);
    }
}

File: sdks/php/src/Crypto/DPoPProof.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Crypto;

/**
 * DPoP proof JWT creation (RFC 9449).
 */
final class DPoPProof
{
    /**
     * Create a DPoP proof JWT.
     *
     * @param string $privateKey 64-byte Ed25519 secret key
     * @param string $method HTTP method (GET, POST, etc.)
     * @param string $url Full URL for the htu claim (must use https://{platformDomain}/path)
     * @param string|null $accessToken If provided, includes ath (access token hash) claim
     * @return string Compact JWS (header.payload.signature)
     */
    public static function create(string $privateKey, string $method, string $url, ?string $accessToken = null): string
    {
        $publicKey = Ed25519::getPublicKey($privateKey);
        $x = Base64Url::encode($publicKey);

        $header = [
            'typ' => 'dpop+jwt',
            'alg' => 'EdDSA',
            'jwk' => [
                'kty' => 'OKP',
                'crv' => 'Ed25519',
                'x' => $x,
            ],
        ];

        $payload = [
            'jti' => self::generateUuid(),
            'htm' => $method,
            'htu' => $url,
            'iat' => time(),
        ];

        if ($accessToken !== null) {
            $ath = hash('sha256', $accessToken, true);
            $payload['ath'] = Base64Url::encode($ath);
        }

        $headerB64 = Base64Url::encode(json_encode($header, JSON_UNESCAPED_SLASHES));
        $payloadB64 = Base64Url::encode(json_encode($payload, JSON_UNESCAPED_SLASHES));
        $signingInput = $headerB64 . '.' . $payloadB64;

        $signature = Ed25519::sign($signingInput, $privateKey);
        $sigB64 = Base64Url::encode($signature);

        return $headerB64 . '.' . $payloadB64 . '.' . $sigB64;
    }

    /**
     * Generate a UUID v4 string.
     */
    private static function generateUuid(): string
    {
        $data = random_bytes(16);
        // Set version to 0100 (v4)
        $data[6] = chr(ord($data[6]) & 0x0f | 0x40);
        // Set variant to 10xx
        $data[8] = chr(ord($data[8]) & 0x3f | 0x80);

        return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
    }
}
